### PR TITLE
New Easier Configuration Option

### PR DESCRIPTION
**Pull request by XBrav**
### Requirements

* Filling out this template is required. Pull Requests without a clear description may be closed at the maintainers' discretion.

### Description

As discussed on Discord, this change is designed to streamline a user's ability to move between versions of Marlin with minimal changes. The concept is to move all declared variables to an #ifndef scenario in Configuration.h and allow the user to enter only their deviations from the default config in a separate file (Currently called User_Config.h). If the parameter is undefined, the stock value will be utilized.

This will also allow users to submit bug reports with all their deviations from the stock config file to help diagnose configuration issues.

### Benefits

As noted above.

### Related Issues

None

### CODE REVIEW

**Comment by XBrav** [comment]
This change only currently impacts Configuration.h, but can be expanded to Configuration_adv.h. It also doesn't handle if you need to undefine stock settings. My current thought is you would need to introduce another file, or additional variables.

**Comment by thinkyhead** [comment]
I would rather not adopt this scheme. It adds a lot more noise to the configuration files and is going to make dealing with configurations sent to us for user support more difficult. I would rather have modified configuration files from users when troubleshooting than just their personal uniquely-ordered differences file.

Also, configurator tools under development are parsing the configuration files to derive dependencies, build data structures, build forms, etc., and the addition of `#ifndef` on every single option makes this work more challenging.

We don’t currently have any difficulty dealing with modified configurations because we can quickly see the changes made by users by using GitHub Desktop or other diff tools that clearly highlight changed lines.

And I think it is important when users share config files publicly that there be a common structure with the documentation included alongside each option. It is also very simple for us to see what version of Marlin and what point in development history the configuration hails from.

For these reasons (and others) I would very much prefer to leave the configuration scheme just as it is.

**Comment by landodragon141** [comment]
I agree with Thinky on this. I think this would be much better handled by an external configurator that changing the internal configuration structure.

**Comment by XBrav** [comment]
The reason I proposed this change is due to the Configuration.h file getting noticeably more cluttered already. It's beginning to feel like there needs to be a Configuration, Configuration_Adv, and Configuration_Adv_Adv.

Most printer configurations will only need a few changes from the master to get running, yet these settings are spread out in the current file. The solution may even be to consider migrating some of the settings over to the config_adv (ie skew correction, ABL bed levelling configurations) so that the Configuration.h remains for core parameters and enabler flags.

**Comment by forkoz** [comment]
meh, I needed changes to config, config_adv and pins. If you want to talk confusing, look at smoothieware. The marlin configs are well documented in the actual files.

**Comment by yucelll** [comment]
I think; Marlin need a GUI,exe.. Users open de GUI and simple basic values change or on/off option and "create" and marlin ready, now open de marlin and compile.

**Comment by comps** [comment]
> The reason I proposed this change is due to the Configuration.h file getting noticeably more cluttered already. It's beginning to feel like there needs to be a Configuration, Configuration_Adv, and Configuration_Adv_Adv.

Then why would you add lots of extra `ifndef` clutter, making navigating the changes even harder?

My standard workflow is to keep my config changes in git and, when I update, I
1) do a diff of my previous base and the would-be-new base to get a basic idea of what config options were added, as I might want to enable some (ie. `S_CURVE_ACCELERATION`)
2) just rebase on top of the latest `bugfix-2.0.x`, resolving any conflicts, enabling any new features I want

This pretty easily ensures that I always have a clean changeset without deprecated options.

A more "regular user" who just wants to enable/disable some features without touching the configs can still do so via `make` arguments, which was IIRC the whole point behind `ENABLED()`/`DISABLED()`.

Alternatively, you can probably maintain a very minimalistic created-from-scratch configuration file symlinked to both `Configuration.h` and `Configuration_adv.h` (so you have only one file), listing only the changed values. If a config option changes, `CONFIGURATION_H_VERSION` and option-specific deprecated warnings in `SanityCheck.h` would tell you if you need to update something.

Finally, there are tools like https://github.com/akaJes/marlin-config which can (AFAIR) apply changes from an old config to a new one.

TL;DR - there are many ways how to maintain your own Marlin configuration, but the proposed change only makes lives more difficult for people that hand-modify the stock `Configuration.h`.

---
If the goal was to make `Configuration.h` into a "defaults" file, with the user just specifying what is changed from those defaults, then IMHO the proper way would be to start moving the default values to `Conditionals_post.h` and leaving them commented by default in `Configuration.h` (and `_adv`).

**Comment by reloxx13** [comment]
from another project i know the user_config method.
and there the user has to use ifdef/ifndef/undef to adjust the default configs. 
that way u/the devs dont need to alter the current configs.

the user config gets loaded AFTER the default config.

this is an example user config from the other project:
https://github.com/arendst/Sonoff-Tasmota/blob/development/sonoff/user_config_override_sample.h

i use it myself and miss it in marlin.

**Comment by Driley142** [comment]
> 
> 
> The reason I proposed this change is due to the Configuration.h file getting **noticeably more cluttered already**. It's beginning to feel like there needs to be a Configuration, Configuration_Adv, and Configuration_Adv_Adv.
> 
> Most printer configurations will only need a few changes from the master to get running, yet these settings are spread out in the current file. The solution may even be to consider migrating some of the settings over to the config_adv (ie skew correction, ABL bed levelling configurations) so that the Configuration.h remains for core parameters and enabler flags.

**Comment by Driley142** [comment]
Yay ! Too much crap in config file. Can't find stuff that matters to my machine, Rosie. She is a workhorse with good OLD Marlin, but has seizures on 2.0 . 
And, a side note, sorry ... I have to get this down in digits now before I forget. Will post it in appropriate something something Bender is great ... why is autocal so restrictive on where it can start from? If the bed sensor says stop, and the DELTA_DIAGONAL_ROD  is good and DELTA_RADIUS is "reasonably close" (not off by 2x, or an order of magnitude) and DELTA_HEIGHT is spot on and DELTA_PRINTABLE_RADIUS is a fairly small % of DELTA_RADIUS, why does it tend to throw a "radius Implausible" ? How do I loosen those constraints ? Sorry, I've just had a hard day trying to figure out whats wrong with my poor sick Rosie. She was fine until someone changed the way we describe the physical shape of a Delta :  (

**Comment by 7eggert** [comment]
When I do a setup, I'll read the config files line by line and adjust the settings. If settings are missing from the file I read, I won't be aware of options I might want to change. That would be OK if my printer would not support these options anyway, but e.g. my Anet A8 does now feature an I²C display, rotary encoder and RGB LEDs. Users might surprise you with the options being applicable to them.

OTOH I had a hard time redefining my pins because the pins_ANET10.h would unconditionally undefine the encoder unless I'd have a graphics display. Also by having an ADC keyboard pin (Home, Back, Encoder-Click == former middle button on the LCD2004), reading the encoder pins was disabled.

TL;DR and ¢¢: The individual example config files may be slimmed down (but not too much) if they reference a complete example config, too.

**Comment by towe96** [comment]
I don't think this proposal would make configuration much easier.
However, I feel like the configuration files should be clustered a bit more clearly.

All the ABL business has become so massive it could probably fill its own `config_abl` file to declutter the main config file.
The same could apply to parts of `configuration_adv`, with more clearly seperated sections for fans, stepper drivers and UI.
